proposals/xxxx-package-manager-branch-support.md

```diff
@@ -0,0 +1,50 @@
+# Package Manager Support for branches
+
+* Proposal: [SE-XXXX](xxxx-package-manager-branch-support.md)
+* Author: [Boris Bügling](https://github.com/neonichu)
+* Review Manager: TBD
+* Status: Discussion
+
+## Introduction
+
+This proposal adds enhancements to the package manifest to support development of packages without strict versioning. This is one of two features, along with "Package Manager Support for Top of Tree development", being proposed to enable use of SwiftPM to develop on "top of tree" of related packages.
+
+## Motivation
+
+The package manager currently supports packages dependencies which are strictly versioned according to semantic versioning. This is how a package's dependencies should be specified when that package is released, but this requirement hinders some development workflows:
+
+- bootstrapping a new package which does not yet have a version at all
+- developing related packages in tandem in between releases, when one package may depend on the latest revision of another, which has not yet been tagged for release
+
+## Proposed solution
+
+As a solution to this problem, we propose to extend the package manifest to allow specifying a branch instead of a version to support revlocked packages and initial bootstrapping.
+
+## Detailed Design
+
+We will introduce a second initializer for `.Package` which takes a branch instead of a version range:
+
+```swift
+import PackageDescription
+
+let package = Package(
+    name: "foo",
+    dependencies: [
+        .Package(url: "http://url/to/bar", branch: "development"),
+    ]
+)
+```
+
+Whenever dependencies are checked out or updated, if a dependency on a package specifies a branch instead of a version, the latest commit on that branch will be checked out for that package. If a dependency on a package specifies a branch instead of a version range, all other dependencies on that package in the current package graph must specify the same branch, and not a version, or an error will be emitted and checking out dependencies will fail.
```

Comment:
I thought this might be confusing, but on the other hand you're right that it contradicts with the previous paragraph and also with the intent of the proposals, because this would mean having to change a bunch of package manifests if a dependency is used by leaf packages.